hikari/files.py

```diff
@@ -470,6 +480,40 @@ async def read(
         async with self.stream(executor=executor) as reader:
             return await reader.read()
 
+    async def save(
+        self,
+        path: typing.Optional[Pathish] = None,
+        *,
+        executor: typing.Optional[concurrent.futures.Executor] = None,
+        force: bool = False,
+    ) -> None:
+        """Save this resource to disk.
+
+        This writes the resource file in chunks, and so does not load
+        the entire resource into memory.
+
+        Parameters
+        ----------
+        path : typing.Optional[Pathish]
+            The path to save this resource to. If this is a string, the
+            path will be relative to the current working directory. If
+            `None`, the resource will be saved as `filename` in
+            the current working directory.
+        executor : typing.Optional[concurrent.futures.Executor]
+            The executor to run in for blocking operations.
+            If `None`, then the default executor is used for
+            the current event loop.
+        force : bool
+            Whether to overwrite an existing file. Defaults to `False`.
+        """
+        loop = asyncio.get_running_loop()
+        path = ensure_path(path or self.filename)
+
+        with await loop.run_in_executor(executor, _open_write_path, path, self.filename, force) as f:
+            async with self.stream(executor=executor) as reader:
```

Comment:
just an idea but a specialisation of this which only makes one thread call to copy the file could prob be implemented specifically on the `File` resource implementation like the following:

```py
import shutil

def _to_write_path(path: pathlib.Path, default_filename: str, force: bool) -> typing.BinaryIO:
    if path.is_dir():
        path = path.joinpath(default_filename)

    if not force and path.exists():
        raise FileExistsError(f"file {path!r} already exists; use `force=True` to overwrite")

    return path


class File(...):
    async def _save_file(self, path, force):
        path = ensure_path(path or self.filename)
        shutil.copy2(self.path, _to_write_path(path, self.filename, force))

    async def save(self, path, executor, force):
        await loop.run_in_executor(executor, self._save_file, path, force)
```


This should optimise by avoiding multiple thread calls but also just better handles copying over the file's permissions and metadata (also shutil tries to use fast copies which avoids reading the data into memory).

Reply:
Implemented!

Thanks for the great idea!